include/onnxruntime/core/framework/int4.h

```diff
@@ -0,0 +1,143 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include <cassert>
+#include <type_traits>
+#include "core/common/common.h"
+#include "core/common/gsl.h"
+
+namespace onnxruntime {
+
+template <bool Signed>
+struct Int4Traits;
+
+template <>
+struct Int4Traits<true> {
+  using UnpackedType = int8_t;
+  static constexpr int8_t min_val = -8;
+  static constexpr int8_t max_val = 7;
+};
+
+template <>
+struct Int4Traits<false> {
+  using UnpackedType = uint8_t;
+  static constexpr uint8_t min_val = 0;
+  static constexpr uint8_t max_val = 15;
+};
+
+/// <summary>
+/// Stores 2 packed 4-bit elements in 1 byte.
+/// </summary>
+/// <typeparam name="Signed">Set to true if signed int4, or false if unsigned uint4.</typeparam>
+template <bool Signed>
+struct Int4x2Base {
+  using UnpackedType = typename Int4Traits<Signed>::UnpackedType;
+  static constexpr UnpackedType min_val = Int4Traits<Signed>::min_val;
+  static constexpr UnpackedType max_val = Int4Traits<Signed>::max_val;
+
+  uint8_t bits_{};
```

Comment:
Do we need to conflate the data type with how it's stored?

Wondering if simply taking a std::byte* (although we use uint8_t in lots of places as std::byte wasn't available at the time) and having a single GetElement(size_t offset) method would be more flexible and slightly more generic.

Reply:
This class already has a `GetElem(size_t index)` method that returns either the lower or upper 4bits.

> std::byte

Sounds good. I'll update to use std::byte instead of uint8_t.

Reply:
Was thinking more along the lines of having a gsl::span<std::byte> where the GetElement could take any index value up to span size x 2, so the Int4x2Base is a more generic Int4Base and we can maybe have an Int4 and UInt4 type to decouple the type from the storage.

Reply:
let's consider this for a future update as @fajin-corp is going to pick up the work from here.